amlib/dsp/stereo_convolution_mac.py

```diff
@@ -0,0 +1,439 @@
+        self._slices = math.ceil(self._tapcount / (clockfrequency / samplerate))  # 4
+        self._size_of_slizes = self._bitwidth * self._slices #how many bits per slice
+        self._samples_per_slice = self._tapcount // self._slices #how many samples per slice
+
+        print(f"Creating {self._slices} slices for {self._tapcount} taps.")
+
+        assert self._tapcount % self._slices == 0, f"Tapcount {self._tapcount} cannot be evenly distributed on {self._slices} slizes."
+
+        taps_fp = taps[:self._tapcount, 0]
+        taps2_fp = taps[:self._tapcount, 1]
+
+        self._taps1_memory = Memory(width=self._size_of_slizes, depth=self._samples_per_slice)
+        self._taps2_memory = Memory(width=self._size_of_slizes, depth=self._samples_per_slice)
+        self._samples1_memory = Memory(width=self._size_of_slizes, depth=self._samples_per_slice)
+        self._samples2_memory = Memory(width=self._size_of_slizes, depth=self._samples_per_slice)
+
+        taps_fp_mod = []
+        taps2_fp_mod = []
+        for i in range(0, self._tapcount, self._slices):
+            val1 = 0
+            val2 = 0
+            for j in range(self._slices):
+                val1 += int(taps_fp[i+j]) << (self._slices - j - 1) * self._bitwidth
+                val2 += int(taps2_fp[i+j]) << (self._slices - j - 1) * self._bitwidth
+            taps_fp_mod.append(val1)
+            taps2_fp_mod.append(val2)
+
+        self._taps1_memory.init = taps_fp_mod
+        self._taps2_memory.init = taps2_fp_mod
+
+    def elaborate(self, platform) -> Module:
+        m = Module()
+
+        taps1_read_port = self._taps1_memory.read_port()
+        taps2_read_port = self._taps2_memory.read_port()
+        samples1_write_port = self._samples1_memory.write_port()
+        samples2_write_port = self._samples2_memory.write_port()
+        samples1_read_port = self._samples1_memory.read_port()
+        samples2_read_port = self._samples2_memory.read_port()
+
+        m.submodules += [taps1_read_port, taps2_read_port, samples1_write_port, samples2_write_port, samples1_read_port, samples2_read_port]
+
+        set1 = Signal()
+        set2 = Signal()
+        output_channels = Signal(2)
+        ix = Signal(range(self._samples_per_slice + 1))
+
+        previous_sample1 = Signal(self._size_of_slizes)
+        previous_sample2 = Signal.like(previous_sample1)
+        current_sample1 = Signal.like(previous_sample1)
+        current_sample2 = Signal.like(previous_sample1)
+        carryover1 = Signal(signed(self._bitwidth))
+        carryover1_2 = Signal.like(carryover1)
+        carryover2 = Signal.like(carryover1)
+        carryover2_2 = Signal.like(carryover1)
+
+        madd_values = Array(Signal(signed(self._bitwidth * 2), name=f"madd_values_{i}") for i in range(self._slices * 2))
+        sumSignalL = Signal(signed(self._bitwidth * 2))
+        sumSignalR = Signal.like(sumSignalL)
+
+        #debugging
```

Comment:
Could you please make that code conditional to a debug flag that can be set by the user?

Reply:
adjusted